test-collateral/generate-policy/src/main.rs

```diff
@@ -0,0 +1,524 @@
+    sinaloa_url: Option<Url>,
+    /// The URL of the Tabasco instance.
+    tabasco_url: Option<Url>,
+    /// The filename of the output policy file.
+    output_policy_file: String,
+    /// The expiry timepoint of the server certificate.  This is measured in
+    /// hours.  Represented as an `i64` as this is what the `chrono` library
+    /// expects.  We check that any value passed as an argument is positive,
+    /// however.
+    certificate_lifetime: i64,
+    /// The data provisioning order.
+    data_provisioning_order: Vec<i32>,
+    /// The streaming provisioning order.
+    streaming_provisioning_order: Vec<i32>,
+    /// The filename of the WASM program.
+    program_binary: String,
+    /// The debug flag.
+    debug: bool,
+    /// Describes the execution strategy (interpretation or JIT) that will be
+    /// used for the computation.
+    execution_strategy: String,
+}
+
+impl Arguments {
+    /// Creates a new `Arguments` structure with all fields set to empty (with
+    /// the `debug` flag set to `false`, and the `certificate_lifetime` field
+    /// set to `0`).
+    #[inline]
+    pub fn new() -> Self {
+        Arguments {
+            certificates: Vec::new(),
+            roles: Vec::new(),
+            sinaloa_url: None,
+            tabasco_url: None,
+            output_policy_file: String::new(),
+            certificate_lifetime: 0,
+            data_provisioning_order: Vec::new(),
+            streaming_provisioning_order: Vec::new(),
+            program_binary: String::new(),
+            debug: false,
+            execution_strategy: String::new(),
+        }
+    }
+}
+
+/// Checks that the string `strategy` matches either "Interpretation" or "JIT",
+/// and if not prints an error message and aborts.
+fn check_execution_strategy(strategy: &str) {
+    if strategy == "Interpretation" || strategy == "JIT" {
+        return;
+    } else {
+        eprintln!("Could not parse execution strategy argument.");
+        exit(1);
+    }
+}
+
+/// Checks that all strings appearing in all vectors in the `roles` argument are
+/// valid Veracruz roles: "ResultReceiver", "DataProvider", or "ProgramProvider".
+fn check_roles(roles: &[Vec<String>]) {
+    if !roles.iter().all(|v| {
+        v.iter()
+            .all(|s| s == "ResultReceiver" || s == "DataProvider" || s == "ProgramProvider")
+    }) {
+        eprintln!("Could not parse the role command line arguments.");
+        exit(1);
+    }
+}
+
+/// Parses the command line options, building a `CommandLineOptions` struct out
+/// of them.  If required options are not present, or if any options are
+/// malformed, this will abort the program.
+fn parse_command_line() -> Arguments {
+    let default_expiry = DEFAULT_CERTIFICATE_EXPIRY.to_string();
+    let default_debug = DEFAULT_DEBUG_STATUS.to_string();
+
+    let matches = App::new(APPLICATION_NAME)
+        .version(VERSION)
+        .author(AUTHORS)
+        .about(ABOUT)
+        .arg(
```

Comment:
Have you considered using structopt to reduce most of this boilerplate?

It uses clap as a backend, so the options are one-to-one
https://github.com/veracruz-project/veracruz/blob/87ef5a9a478ccc9b0a9514a54fe92fdc2695ca17/durango-cli/src/main.rs#L67

Reply:
No, I've never used `structopt` before.  I'll take a look if that's what you are using already in another CLI tool.

Reply:
It's what I was using for the other CLIs, though I can change to clap if we prefer that.

It wraps clap in macro magic, leveraging rustdocs comments/type-introspection. So something like this:
``` rust
Arg::with_name("certificate")
                .short("c")
                .long("certificate")
                .value_name("FILE")
                .help("The filename of a cryptographic certificate identifying a computation participant.")
                .required(true)
                .multiple(true),
```

Can be written like this:
``` rust
/// The filename of a cryptographic certificate identifying a computation participant.
#[structopt(short, long, value_name="FILE", multiple=true)]
certificate: Vec<String>,
```

Pretty nifty IMO

Reply:
Ah, I did not realize `parse` was a structopt-specific extension. This may (or may not) let you simplify some of the parsing logic:
https://docs.rs/structopt/0.3.21/structopt/index.html#custom-string-parsers

Reply:
I think, after thinking more, that I'm going to leave the code as it is now, given it's already written.

Reply:
Sounds like a good enough reason to me. Mind if I adopt structopt here in a later PR for consistency with the other CLI interfaces?

Reply:
Yeah, feel free!